pkg/etcd-backup-operator/controller/s3backup.go

```diff
@@ -0,0 +1,229 @@
+		logrus.Warning(msg)
+		return fmt.Errorf(msg)
+	}
+	member, rev := getMemberWithMaxRev(pods, nil)
+	if member == nil {
+		msg := "no reachable member"
+		logrus.Warning(msg)
+		return fmt.Errorf(msg)
+	}
+
+	logrus.Infof("saving backup for cluster %s ...", s.clusterName)
+	if err := s.writeSnap(member, rev); err != nil {
+		msg := fmt.Sprintf("failed to save backup for cluster %s: (%v)", s.clusterName, err)
+		logrus.Warning(msg)
+		return fmt.Errorf(msg)
+	}
+	logrus.Infof("saving backup for cluster %s succeeded", s.clusterName)
+	return nil
+}
+
+func (s *s3Backup) writeSnap(m *etcdutil.Member, rev int64) error {
+	cfg := clientv3.Config{
+		Endpoints:   []string{m.ClientURL()},
+		DialTimeout: constants.DefaultDialTimeout,
+		TLS:         nil,
+	}
+	etcdcli, err := clientv3.New(cfg)
+	if err != nil {
+		return fmt.Errorf("failed to create etcd client: (%v)", err)
+	}
+	defer etcdcli.Close()
+
+	ctx, cancel := context.WithTimeout(context.Background(), constants.DefaultRequestTimeout)
+	resp, err := etcdcli.Maintenance.Status(ctx, m.ClientURL())
+	cancel()
+	if err != nil {
+		return fmt.Errorf("failed to get member %v status: (%v)", m.ClientURL(), err)
+	}
+
+	ctx, cancel = context.WithTimeout(context.Background(), constants.DefaultSnapshotTimeout)
+	defer cancel()
+	rc, err := etcdcli.Maintenance.Snapshot(ctx)
+	defer rc.Close()
+	if err != nil {
+		return fmt.Errorf("failed to receive snapshot: (%v)", err)
+	}
+	err = s.saveToS3(resp.Version, rev, rc)
+	if err != nil {
+		return fmt.Errorf("failed to save snapshot: (%v)", err)
+	}
+	return nil
+}
+
+func (s *s3Backup) saveToS3(version string, rev int64, rc io.ReadCloser) error {
+	secret, err := s.kubecli.Core().Secrets(s.namespace).Get(s.awsSecret, metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to retrieve secret %v: (%v)", s.awsSecret, err)
+	}
+	s3u, err := newS3Uploader(s.s3bucket, s.prefix, secret)
+	if err != nil {
+		return fmt.Errorf("failed to create s3 uploader: (%v)", err)
+	}
+	key := path.Join(s.prefix, makeBackupName(version, rev))
+	ui := &s3manager.UploadInput{
+		Bucket: aws.String(s.s3bucket),
+		Key:    aws.String(key),
+		Body:   rc,
+	}
+	_, err = s3u.Upload(ui)
+	if err != nil {
+		return fmt.Errorf("failed to upload snapshot to s3: (%v)", err)
+	}
+	return nil
+}
+
+func makeBackupName(ver string, rev int64) string {
+	return fmt.Sprintf("%s_%016x_%s", ver, rev, backupFilenameSuffix)
+}
+
+// NewS3Uploader returns a Uploader that is configured with proper aws config and credentials derived from Kubernetes secret.
+func newS3Uploader(bucket, prefix string, secret *v1.Secret) (*s3manager.Uploader, error) {
+	so, err := newAWSConfig(secret, prefix)
+	if err != nil {
+		return nil, fmt.Errorf("failed to setup aws config: (%v)", err)
+	}
+	return newFromSessionOpt(bucket, prefix, *so)
+}
+
+func newAWSConfig(secret *v1.Secret, prefix string) (*session.Options, error) {
+	dir := filepath.Join(operatorRoot, "aws", prefix)
```

Comment:
If operatorRoot is only used for storing temporary credentials, I would suggest we just store it under "/tmp" and make tempdir there.

Reply:
Also this codepath should be refactored and reused.
This is unnecessary duplicate.